### DISCUSSION TITLE
Hello! :)

### DISCUSSION THREAD

**Discussion by Chris-Xavier** [community/community#138059] [Other Feature Feedback, Questions, & Ideas] [answered]
### Select Topic Area

Question

### Body

So, 
My name is Christian but I go by Chris.
I am pretty new to Github and am looking for ANY advice to help me get started.
I am currently a student at a tech program and am learning the fundamentals for programming!
My dream goal is to create a social media application that will dominate the social networking industry 
and am looking for any advice for how I could get started with that. 👍🏻 
I am in the planning phase for the application and have not started coding one bit. 
Anyways, I really want to learn from you guys and am looking to consume almost any information you guys could throw at me that will help me completely! Thank you!

**Comment by Shriharsh-Deshmukh** [comment]
Welcome to GitHub, Chris! You're off to a great start by being eager to learn. Here’s a tech-focused roadmap to help you get started on your journey toward building that dominating social media application:

### 1. **Version Control with Git:**
   - First, get comfortable with Git. Learn basic commands like `git clone`, `git commit`, `git push`, and `git pull`. This will be essential for version control as you start coding.
   - Try to contribute to open-source projects on GitHub. It will help you learn collaboration and how large-scale projects are managed.

### 2. **Project Planning:**
   - Before diving into coding, create a **well-structured project roadmap**. Break down your social media app into modules (e.g., user authentication, profile management, posts, messaging).
   - Use GitHub Projects to manage tasks, track issues, and collaborate with others. Start by setting up a **Kanban board** with columns like "To Do," "In Progress," and "Done."

### 3. **Learn the Tech Stack:**
   - Choose a tech stack that aligns with your goals. For a social media app, you might consider:
     - **Frontend:** React.js or Angular for dynamic and responsive UI.
     - **Backend:** Node.js with Express.js, or Django (Python) for building RESTful APIs.
     - **Database:** PostgreSQL or MongoDB for managing user data and posts.
     - **Authentication:** OAuth or JWT for user authentication and authorization.
   - Practice building smaller projects using this stack to solidify your knowledge.

### 4. **Start Small and Scale:**
   - Begin with an MVP (Minimum Viable Product). Build the core features first (e.g., user registration, login, posting). This will allow you to iterate and improve based on user feedback.
   - Use **GitHub Issues** to track bugs and feature requests as you build and scale your app.

### 5. **Networking and Learning:**
   - Follow **GitHub Discussions** and other developers in your field. Engage in coding communities like Stack Overflow, Reddit, and Twitter to stay updated with the latest trends.
   - Consider joining a **Hackathon** or collaborating on projects with others to sharpen your skills.

### 6. **Deployment:**
   - Once you're ready, explore **CI/CD pipelines** (Continuous Integration/Continuous Deployment) using GitHub Actions. This will automate testing, building, and deploying your application.
   - Look into cloud platforms like **Heroku**, **AWS**, or **DigitalOcean** for hosting your app.

### 7. **Security and Optimization:**
   - Implement security best practices, such as input validation and encryption for sensitive data.
   - Optimize your application for performance, especially if you're targeting a large user base.

Keep learning and stay persistent. GitHub and its community can be your best resources as you build your app. Welcome to the world of code, and good luck with your journey to creating the next big social media platform!

**Reply by novium2** [reply]
Hi @Shriharsh-Deshmukh
We’ve clarified our stance on using generative AI tools like ChatGPT within our Community via [this announcement](https://github.com/orgs/community/discussions/69006). Please review the guidelines to ensure your post meets them as failure to adhere to those rules can result in action taken by our moderator team.  You can read our updated [Code of Conduct](https://github.com/community/community/blob/baa5cd9020ef3f2bbf86dca1191d9c90f9b0822d/CODE_OF_CONDUCT.md#reasonable-use-of-ai-generated-content) and the [announcement](https://github.com/orgs/community/discussions/69006) for more details. Thank you for helping us maintain an authentic and beneficial space for everyone.

**Comment by Sin-cy** [comment, marked as answer]
Heyyyy. I struggled quite a bit at the start as well so here are the Things that I wished I knew when I first started and some tips along the way.

First one is understanding what Git is because to properly use Github, you gotta understand git first imo. I'll just explain some things that you will absolutely need if you're gonna get into Github. 
Github is like a hub lul, gotta understand how people use Github. What they use it for. 
Then go on to understanding, how to upload or push files from your system to this Github or vice versa. Its basically like a website that all devs come to. 
Either to upload/share your project which are like files and in those files are many more sub directories/folders. That sort of thing. They are basically code files that you write on your system and then upload it up here into a repository or know as folders you could say.

You can download these folders from anyone as long as they have status set to Public. You do this by doing running a command 
`git clone "and the link of their repository"`
This is like one of the first command u'll learn when starting out. This is what I meant by you gotta Understand what Git is first.
Before this you might want to go research about how to Install Git on your system. Once you do that. 
Then comes the version control part. Sounds confusing but its just folders with multiple version of your code pushed onto Github repositories which are like folders.
Version control works by running different types of commands. So in your terminal/shell of your system.

### Basics
These are going to be the most used commands at first. ( You gotta understand what these command does )
`git init`
`git clone`
`git add .`
`git status`
`git commit -m "describe what you did with the file in short"`
`git push`
`git pull`

Its confusing to get into at first, but stick to it and don't go looking for alternatives that makes version control easier. You can do that later. Just understand the commands first and how they work. 

Then once you get some of the basics commands. Then you go and start learning about
- Remote Repository and Local Repository or Repos ( what are the differences between the two , because some commands work well with remote and some with local )
-  what is git branches, how branches are created, how branches are updated , how to change branch names and how they link up with remote and local repos ( such as : Branch A has a remote repo set but Branch B may not have a remote repo set). Learn the differences between them.
- Then Start looking at commands like 
`git branch` which will lead you to other git branch commands that links up with `git switch` for switching your branch

Most importantly, understand 

- what are commit hashes , why they exist. How to make commits, how and when should you undo commits or rename commits.
- also I can't stress this enough when I got started. I didn't know where to look because there is so much going on sometimes.
- Just learn where to look, because if you look at the right places then you'll go faster. ( such as which commands can help you look, understand and find what you need faster )

### Merge and Rebase
One Important thing before going into this is to understand that. 
**When doing Merge or Rebase, you want to be on the branch that you are looking to merge/rebase.**
 You'll only understand it once you get the `git branch` part. The Same goes for the rest of the things Im going to say below. 

- Merging  is probably the next up. Understand the merging concepts such as 
`git merge` command and thier limitations. When to use and when not use. How to use etc.

- Rebase is also very important, but you might want to understand branch and merging fully before getting into it. 
`git rebase` command
There are a lot of do's and don't with git rebase and git merge but I'd recommend you go researching it and experiment it with your own test projects. Try merging and rebasing and running all the main git commands, see if you understand whats going on. Aaagain doing and putting those commands into action is probably the best way to learn. 

Some other commands you might find useful later along the learning path
`git log` 
`git log --oneline --graph --all` or `git --no-pager log --oneline --graph --all`
`git stash` which will lead you to `git worktrees` (very useful btw)
`git ammend`
`git reset`
`git revert`
etc....

Then once you get everything you can go and mess with aliases/shortcuts for your git command to type them out shorter or even inside your git configuration settings.

This should help you get started into Git and Github :) 💪
